joomla/joomla-cms
# [4.0] Add table rename to database schema check

Pull Request for [https://github.com/joomla/joomla-cms/issues/29279#issuecomment-636364871](https://github.com/joomla/joomla-cms/issues/29279#issuecomment-636364871) (partly).

### Summary of Changes

This Pull Request (PR) adds a check for table rename to the database schema check and fix so that such statements in update SQL scripts will not be ignored anymore by the database check and fix.

Currently without this PR the database schema check and fix ignores the following schema change:
- For MySQL
https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_admin/sql/updates/mysql/4.0.0-2020-05-21.sql#L2
- For PostgreSQL
https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_admin/sql/updates/postgresql/4.0.0-2020-05-21.sql#L2

### Testing Instructions

There are 2 tests to be done:
1. Reproduce the issue without the patch of this PR applied.
2. Check that the issue is fixed with the patch of this PR applied.

It needs tests with MySQL and with PostgreSQL databases.

**_Testers please report back which kind of database you have used for the test. If you have both, please test both._**

We can better count then if we have enough tests for both kinds of databases.

Currently we have:
- One successful test with MySQL.
- No test yet with PostgreSQL.

#### Test 1: Reproduce issue

1. On a clean current 4.0-dev or latest nightly or 4.0 Beta 1, login to backend and go to "System - Information - Database"
Result: There are no database errors shown for the CMS.
2. Move the mouse over the green badge of the CMS to see the details in a pop-up.
Result:
![database-no-errors-pr-not-applied](https://user-images.githubusercontent.com/7413183/83404029-207e5580-a40a-11ea-8e57-1db8b34a9944.png)
3. Note the numbers of checked and skipped database changes for later test "Test 2".
4. Using a tool like e.g. phpMyAdmin or phpPgAdmin (depending on your database type), revert the renaming of the table in the statement mentioned in the Summary of Changes above:
- MySQL:
```
RENAME TABLE `#__history` TO `#__ucm_history`;
```
- PostgreSQL:
```
ALTER TABLE "#__history" RENAME TO "#__ucm_history";
```
(replace `#__` by your database prefix).

5. Go again to "System - Information - Database", or of you still are on that page, close all alerts and refresh the page:
Result: There are database errors shown, but not about a missing table, only about wrong columns.
![database-errors-pr-not-applied-table-renamed-back-to-old](https://user-images.githubusercontent.com/7413183/83404912-bff01800-a40b-11ea-9a48-7a303aad5765.png)
6. Close all alerts.
7. Select the check bow for the row with the CMS and use the "Update Structure" button (formerly known as "Fix").
Result: Same as before in step 5, the errors can't be fixed.

#### Test 2: Test the fix

You have to start again with a clean installation, i.e. not use the one from the test before.

1. Apply the patch of this PR on a clean current 4.0-dev or latest nightly or 4.0 Beta 1 installation.
2. Login to backend and go to "System - Information - Database"
Result: There are no database errors shown for the CMS.
3. Move the mouse over the green badge of the CMS to see the details in a pop-up.
Result:
![database-no-errors-pr-applied](https://user-images.githubusercontent.com/7413183/83405284-7ce27480-a40c-11ea-98c7-30db78b86210.png)
4. Compare the numbers of checked and skipped database changes with those noted in step 3 of the previous test "Test 1".
Result: One more change is checked and one less is skipped.
5. Using a tool like e.g. phpMyAdmin or phpPgAdmin (depending on your database type), revert the renaming of the table in the statement mentioned in the Summary of Changes above:
- MySQL:
```
RENAME TABLE `#__history` TO `#__ucm_history`;
```
- PostgreSQL:
```
ALTER TABLE "#__history" RENAME TO "#__ucm_history";
```
(replace `#__` by your database prefix).

6. Go again to "System - Information - Database", or of you still are on that page, close all alerts and refresh the page:
Result: There are database errors shown about a missing table and about wrong columns.
![database-errors-pr-applied-table-renamed-back-to-old](https://user-images.githubusercontent.com/7413183/83405601-1e69c600-a40d-11ea-9594-5a5ffe9d4bc4.png)
7. Close all alerts.
8. Select the check bow for the row with the CMS and use the "Update Structure" button (formerly known as "Fix").
Result: Same as steps 2 and 3, no errors, database is up to date. 

### Expected result

Table rename statements in update SQL scripts are handled by the database schema check and fix.

### Actual result

Table rename statements in update SQL scripts are ignored by the database schema check and fix.

### Documentation Changes Required

None, since the way how the database check and fix works is not documented anywhere except in the heads of a few people.

### Additional information

Same as the existing checks for renaming columns, the check for renaming tables added by this PR does not check for the old name, i.e. if the table or column has a name different to both the old and the new name, the check doesn't work. To fix this is out of the cope of this PR.
